Library/Dependencies/napi/source/js_native_api_JavaScriptCore.cc

```diff
@@ -1,353 +1,911 @@
-  return napi_ok;
+template<int functionIndex> JSValueRef JSCStaticMethod(JSContextRef ctx, JSObjectRef function, JSObjectRef object, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception) {
+  // vf table
+  auto iter = FunctionTables.find(object);
+  assert(iter != FunctionTables.end());
+  
+  const FunctionTable& table = *iter->second;
+  const FunctionTableEntry& entry = table.table[functionIndex];
+  
+  CallbackInfo callbackInfo;
+  callbackInfo.argc = argumentCount;
+  auto jsValueRefs = const_cast<JSValueRef*>(arguments);
+  auto jsValues = const_cast<OpaqueJSValue**>(jsValueRefs);
+  callbackInfo.argv = reinterpret_cast<napi_value*>(jsValues);
+  callbackInfo.thisArg = reinterpret_cast<napi_value>(object);
+  callbackInfo.data = entry.data;
+  callbackInfo.isConstructCall = false;
+  callbackInfo.newTarget = 0;
+
+  napi_value retValue = entry.cb(table.env, reinterpret_cast<napi_callback_info>(&callbackInfo));
+  
+  return reinterpret_cast<JSValueRef>(retValue);
 }
 
-napi_status napi_get_array_length(napi_env env,
-                                  napi_value v,
-                                  uint32_t* result) {
+void JSCFinalize(JSObjectRef object) {
+  // todo
+  assert(0);
+}
 
-  return napi_ok;
+JSObjectRef JSCCallAsConstructor(JSContextRef ctx, JSObjectRef constructor, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception) {
+
+  auto iter = constructorCB.find(constructor);
+  assert(iter != constructorCB.end());
+  ClassTable *cbInfo = iter->second;
+  
+  
+  assert(JSObjectIsConstructor(ctx, constructor));
+  OpaqueJSValue* jsValue = const_cast<OpaqueJSValue*>(JSObjectMake(ctx, cbInfo->classRef, nullptr));
+  napi_value value = reinterpret_cast<napi_value>(jsValue);
+  
+  FunctionTables[jsValue] = &cbInfo->FunctionTable;
+  
+  // callback will wrap the pointer itself
+  CallbackInfo callbackInfo;
+  callbackInfo.argc = argumentCount;
+  auto jsValueRefs = const_cast<JSValueRef*>(arguments);
+  auto jsValues = const_cast<OpaqueJSValue**>(jsValueRefs);
+  callbackInfo.argv = reinterpret_cast<napi_value*>(jsValues);
+  callbackInfo.thisArg = value;
+  callbackInfo.data = cbInfo->data;
+  callbackInfo.isConstructCall = true;
+  cbInfo->cb(cbInfo->env, reinterpret_cast<napi_callback_info>(&callbackInfo));
+  
+  return jsValue;
 }
 
-napi_status napi_get_arraybuffer_info(napi_env env,
-                                      napi_value arraybuffer,
-                                      void** data,
-                                      size_t* byte_length) {
-  return napi_ok;
+bool JSCSetProperty(JSContextRef ctx, JSObjectRef object, JSStringRef propertyName, JSValueRef value, JSValueRef* exception) {
+  assert(0);
+  return false;
 }
 
+JSValueRef JSCGetProperty(JSContextRef ctx, JSObjectRef object, JSStringRef propertyName, JSValueRef* exception) {
+  auto iter = FunctionTables.find(object);
+  assert(iter != FunctionTables.end());
+  
+  
+  char propertyStr[1024];
```

Comment:
Seems like you should call JSStringGetMaximumUTF8CStringSize instead of hard coded size

Reply:
Fixed